lms/djangoapps/grades/models.py

```diff
@@ -0,0 +1,166 @@
+
+from coursewarehistoryextended.fields import UnsignedBigIntAutoField
+from xmodule_django.models import BlockTypeKeyField, CourseKeyField
+
+import json
+
+
+class VisibleBlocksModel(models.Model):
+    """
+    A django model used to track the state of a set of visible blocks under a given subsection at the time they are
+    used for grade calculation.
+
+    This state is represented using an array of serialized BlockRecords, stored in the blocks_json field. A
+    hash of this json array is used for lookup purposes.
+    """
+    _blocks_json = models.TextField(db_column="blocks_json")
+    hashed = models.CharField(max_length=255, primary_key=True)
+
+    def __init__(self, *args, **kwargs):
+        """
+        Creates a new VisibleBlocksModel. Keyword argument 'blocks' should be an array of BlockRecords. KeyError will
+        be raised if this argument is not present.
+        """
+        blocks = kwargs.pop('blocks')
+        super(VisibleBlocksModel, self).__init__(*args, **kwargs)
+        self._blocks_json = json.dumps(
+            [
+                block.to_json()
+                for block in blocks
+            ]
+        )
+        self.hashed = hash(self._blocks_json)  # TODO: better hash algo here?
+
+    @property
+    def blocks(self):
+        """
+        Returns the blocks_json data stored on this model as an array of BlockRecords. If for some reason block_json
+        is not parsable, the json error will bubble up.
+        """
+        block_dicts = json.loads(self._blocks_json)
+        return [BlockRecord(data) for data in block_dicts]
+
+    @blocks.setter
+    def blocks(self, value):
+        """
+        Not implemented, as VisibleBlocks instances are intended to be write-once and not change after creation.
+        """
+        raise NotImplementedError(
+            "Property 'blocks' cannot be modified on an existing VisibleBlocksModel. Create a new instance."
+        )
+
+
+class PersistentSubsectionGradeModel(TimeStampedModel):
+    """
+    A django model tracking persistent grades at the subsection level.
+
+    TODO: intended query patterns are fluid at the moment, document them here as they're firmed up.
+    """
+
+    class Meta(object):
+        index_together = [
+            # TODO: nail down indices as we flesh out the API layer
+        ]
+
+        unique_together = (('user_id', 'course_id', 'block_type', 'block_id'))
+
+    id = UnsignedBigIntAutoField(primary_key=True)  # pylint: disable=invalid-name
+    subtree_edited_date = models.DateTimeField('last content edit timestamp')
+    user_id = models.CharField(max_length=255)
+    earned_all = models.IntegerField()
+    possible_all = models.IntegerField()
+    earned_graded = models.IntegerField()
+    possible_graded = models.IntegerField()
+
+    # TODO: Make sure this matches up w/ what Nimisha's exposing in her PR
+    course_version = models.CharField('guid of latest course version', max_length=255)
```

Comment:
This particular field was originally going to be used as a way to re-construct the course at the time the student received their grade. That work has been deferred though; the primary value of this as it currently stands is providing data for debugging purposes.